codice/ddf
# DDF-6152 Added standalone solr gazetteer

#### What does this PR do?
Adds a standalone solr gazetteer
[Good writeup in comment below](https://github.com/codice/ddf/pull/6153#issuecomment-655136232)
#### Who is reviewing it? 
@pklinef
@lambeaux

#### How should this be tested?

1. Build PR
2. Unzip and start distribution
2a. `./bin/ddf`
2b. `profile:install standard`
3. Install standalone solr gazetteer feature, `feature:install catalog-solr-offline-gazetteer`
4. If you don't have a working version of catalog-ui , build and then install (ddf-ui master *should* be working with codice/ddf/master (2.26.0-SNAPSHOT))
4a. install catalog-ui, `feature:repo-add mvn:org.codice.ddf.search/intrigue-ui-app/3.7.0-SNAPSHOT/xml/features && feature:repo-add mvn:org.codice.ddf.search/ui-frontend/3.7.0-SNAPSHOT/xml/features && feature:install catalog-ui-app ui-frontend`
5. Go to admin settings and configure catalog-ui to use offline gazetteer 
5a. https://localhost:8993/admin
5b. click `system` tab on top of page
5c. search for `catalog ui`
5d. edit config and **uncheck** the box labeled `Use online gazetteer service`

6. Ingest gazetteer data
6a. `feature:install offline-gazetteer-index`
6b. It may take 5-20 minutes for all ~70k records to install
7. Ensure the data has been picked up by the catalog plugin and placed into the core
7a. visit https://localhost:8994/solr. 
7b. Select `gazetteer` core > query > run the default query and ensure there is ~70k items in the core
8. Build the suggester  by running the `offline-solr-gazetteer:build-suggester-index` command 
8a. You can also build the suggester from the solr dashboard (https://localhost:8994/solr). Go to `gazetteer` core >  query > change request handler (qt) to `/gazetteer` and inside Raw Query Parameters place `suggest.build=true&suggest.dictionary=gazetteerSuggest`
9. Go to intrigue https://localhost:8993/search/catalog
9a. Use the gazetteer search in top right corner of 3d map and ensure searching works
10. Use the advanced search of intrigue to search for gazetteer records in the catalog
10a. New Search > (triple dot overflow menu) > Advanced Search > Attribute= `metacard-tags` and value= `gazetteer`
10b. Modify one item to have a new title with easily searchable text, eg `itworks`.
10c. Rebuild the suggester (See step 8 and 8a)
10d. Use gazetteer search (See step 9) and ensure you can find `itworks` with gazetteer
11. Use the advanced search (see step 10a) to find a gazetteer record and delete it
11a. Rebuild the suggester (see steps 8 and 8a)
11b. ensure that item does not appear in the gazetteer.
12. use the `offline-solr-gazetteer:removeall` command and wait to make sure it removes all items.
12a. wait at least 30 seconds for the index to flush
12b. rebuild the suggester index (see step 8)
12c. ensure no records are in gazetteer by searching with gazetteer
12d. use the `offline-solr-gazetteer:synccatalog` to sync the catalog with the gazetteer core. 
12e. wait at least 30 seconds for the index to flush
12f. rebuild the suggester index (see step 8)
12g. ensure gazetteer in intrigue works and gets results. 

#### Any background context you want to provide?

#### What are the relevant tickets?
Fixes: #6152

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
